### ISSUE TITLE
More structured JSON logging

### ISSUE THREAD

**Issue by almostearthling** [almostearthling/whenever#4]
The logging system should be able to produce JSON messages where the structure referenced in [common.rs](https://github.com/almostearthling/whenever/blob/765cd94851bd8fced2262517c9ce5b91f22bfbae/src/common.rs#L6) is explicitly rendered with separate fields. The emitted JSON output should thus have the following form:

```json
{
    "application": "whenever",
    "time": "2023-10-20T18:25:43.511Z",
    "level": "info",
    "context": {
        "emitter": "CONDITION_REGISTRY",
        "item": "tick",
        "item_id": null
    },
    "message_type": {
        "where": "PROC",
        "status": "MSG"
    },
    "message": "some human readable message"
}
```

(or something similar), in order to let a wrapper be able to better interpret log messages and subsequently react, for example by showing a notification when certain events are logged.

A possible solution is probably to refactor the specific **whenever** [log](https://github.com/almostearthling/whenever/blob/765cd94851bd8fced2262517c9ce5b91f22bfbae/src/common.rs#L249) function so that it accepts all the appropriate arguments, and to correctly reference it in the code. This issue should be solved without actually changing the current textual log format: in fact all of the above described fields can easily be combined into the current format by means of the `format!` macro.

**Comment by almostearthling** [comment]
This format might be better:

```json
{
    "application": "whenever",
    "time": "2023-10-20T18:25:43.511Z",
    "severity": "info",
    "context": {
        "emitter": "CONDITION_REGISTRY",
        "action": "tick",
        "item": "ConditionName",
        "item_id": 12
    },
    "message_type": {
        "where": "PROC",
        "status": "MSG"
    },
    "message": "some human readable message"
}
```

where:

* _action_ can be `null`
* _item_ and _item_id_ should either be both `null` or both evaluated (respectively: a string and an integer)

in order to avoid ambiguity. The logging function should implement a format that avoids writing `null` values to either the console or a non-JSON file. Possible signature of the new `common::logging::log` function:

```rust
    pub fn log(
        severity: LogType,
        emitter: &str,
        action: Option<&str>,
        item: Option<&str>,
        item_id: Option<u64>,
        where: &str,
        status: &str,
        message: &str,
    ) {
        match severity {
            // ...
        }
    }
```

A further improvement would be to use enums for _where_ and _status_; not sure about _action_: however `pub const &str` can however be used for this parameter.

**Comment by almostearthling** [comment]
Some small changes to avoid syntax errors and to comply with the fact that `item` and `item_id` must be both either defined or undefined: the new `log()` function is defined as

```rust
    pub fn log(
        severity: LogType,
        emitter: &str,
        action: &str,               // there should always be an action
        item: Option<(&str, i64)>,  // this is to define an item when needed
        when: &str,                 // where is a reserved word
        status: &str,
        message: &str,
    ) {
            // ...
        }
    }
```

As seen above, the string for _action_ must always be provided.

**Comment by almostearthling** [comment]
Due to the way the `log` facade exposes its macros, that is, only a string can be passed as payload, the structure of the log message had to be changed to this:

```json
{
    "header": {
        "application": "whenever",
        "level": "TRACE",
        "time": "2023-11-04T11:17:25.257970"
    },
    "contents": {
        "context": {
            "action": "scheduler_tick",
            "emitter": "MAIN",
            "item": null,
            "item_id": null
        },
        "message": "condition Cond_TIME tested with no outcome (tasks not executed)",
        "message_type": {
            "status": "MSG",
            "when": "PROC"
        }
    }
}
```

(values differ from the example above, as well as field order, because this record is taken from an actual **whenever** session). This has been implemented in 9b9de2b. It is actually more complex, but at least there is no loss of information in comparison to the original goal: being in pre-release state this can be considered acceptable and stabilized.

Not actually a regression, because this feature is not yet available in the main branch, but _registry_ actions on specific items (actually, on _conditions_) cannot reference the item _id_ in the log payload: this causes a semi-deadlock due to double acquisition of a mutex that causes logging to hang for the specific action: see [commented out code](https://github.com/almostearthling/whenever/blob/9b9de2be56f14ba090a9f1a94927c9b286021d22/src/condition/registry.rs#L420) for an example. The _registries_ definitely need a `get_id(name: &str)`.
